In [1]:
import os
import yaml
import getpass
import sqlalchemy
import numpy as np
import pandas as pd
from scipy import stats
import geopandas as gpd
import plotly.express as px
from dotenv import load_dotenv
from shapely.geometry import Point
from sqlalchemy import create_engine

In [2]:
load_dotenv()
password = os.getenv("DB_PASSWORD")

In [3]:
engine = create_engine(f"mysql+mysqlconnector://root:{password}@localhost/river_risk_index")

In [4]:
# 1. Load Configuration
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)

# Get file path from config
file_path = config["output_data"]["file3"]
# 2. Load Parquet Data
df = pd.read_parquet(file_path)

In [ ]:
# Display columns to know exactly what to map
print("\nAvailable Columns:")
print(df.columns.tolist())


Available Columns:
['OBJECTID', 'lat', 'lng', 'ocean', 'setting', 'sampling_method', 'mesh_size_mm', 'microplastics_measurement', 'unit', 'concentration_class', 'sample_date', 'x', 'y', 'sample_datetime']


In [6]:
# 3. Clean & Filter
# Keep only 'Ocean water' samples (ignore beaches, rivers, sediment)
df = df[df['setting'] == 'Ocean water'].copy()
print(f"✅ Filtered for 'Ocean water'. Rows: {len(df)}")

# Extract Year from sample_date
df['year'] = pd.to_datetime(df['sample_date']).dt.year

# Select and Rename columns for SQL
df_sql = df[[
    'year', 
    'lat', 
    'lng', 
    'microplastics_measurement'  # This is the concentration
]].copy()

# Rename for clarity in SQL
df_sql = df_sql.rename(columns={'microplastics_measurement': 'concentration'})

# Ensure numeric types
df_sql['lat'] = pd.to_numeric(df_sql['lat'], errors='coerce')
df_sql['lng'] = pd.to_numeric(df_sql['lng'], errors='coerce')
df_sql['concentration'] = pd.to_numeric(df_sql['concentration'], errors='coerce')
df_sql['year'] = df_sql['year'].astype('Int64') # Handle nullable integers

✅ Filtered for 'Ocean water'. Rows: 15534


In [ ]:
print("Data Quality Check (Before Dropping):")
print(f"Total rows after filtering 'Ocean water': {len(df_sql)}")

# Check for missing values in critical columns
missing_summary = df_sql[['year', 'lat', 'lng', 'concentration']].isnull().sum()
print("\nMissing Values per Column:")
print(missing_summary)


Data Quality Check (Before Dropping):
Total rows after filtering 'Ocean water': 15534

Missing Values per Column:
year             0
lat              0
lng              0
concentration    0
dtype: int64


In [8]:
"""# Upload to MySQL
df_sql.to_sql('observed_plastic', con=engine, if_exists='replace', index=False)

print(f"SUCCESS: Uploaded {len(df_sql)} rows to 'observed_plastic'")
print(f"Year Range: {df_sql['year'].min()} - {df_sql['year'].max()}")"""

'# Upload to MySQL\ndf_sql.to_sql(\'observed_plastic\', con=engine, if_exists=\'replace\', index=False)\n\nprint(f"SUCCESS: Uploaded {len(df_sql)} rows to \'observed_plastic\'")\nprint(f"Year Range: {df_sql[\'year\'].min()} - {df_sql[\'year\'].max()}")'

In [ ]:
# 1. Load data
print("Loading data from SQL...")
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)
emission_df = pd.read_sql("SELECT lat, lon, emission_tons_year FROM emission_points", engine)

# 2. Convert to GeoDataFrames
print("Converting to GeoDataFrames...")
plastic_gdf = gpd.GeoDataFrame(plastic_df, geometry=gpd.points_from_xy(plastic_df['lng'], plastic_df['lat']), crs="EPSG:4326")
emission_gdf = gpd.GeoDataFrame(emission_df, geometry=gpd.points_from_xy(emission_df['lon'], emission_df['lat']), crs="EPSG:4326")

# 3. Spatial Join: Find Nearest Emission Point
# We use a 'nearest' join. Note: This requires SciPy. 
# If this fails, we can use a simpler distance calculation on a subset.
print("Calculating nearest neighbors (this may take a moment)...")

# Method: Calculate distance to all emissions? No, too slow.
# Better Method: Use sjoin_nearest (GeoPandas 1.0+)
try:
    joined_gdf = gpd.sjoin_nearest(plastic_gdf, emission_gdf, how="left", distance_col="dist_meters")
    
    # Convert meters to KM
    joined_gdf['min_distance_km'] = joined_gdf['dist_meters'] / 1000
    
    # Keep relevant columns
    result_df = joined_gdf[['concentration', 'min_distance_km']].copy()
    
    # Categorize
    def categorize(dist):
        if dist <= 50: return 'Near River (<50km)'
        elif dist <= 200: return 'Regional (50-200km)'
        else: return 'Open Ocean / Far (>200km)'
    
    result_df['proximity_zone'] = result_df['min_distance_km'].apply(categorize)
    
    # 4. Aggregate and Upload Results
    summary_df = result_df.groupby('proximity_zone').agg(
        sample_count=('concentration', 'count'),
        avg_density=('concentration', 'mean'),
        max_hotspot=('concentration', 'max')
    ).reset_index()
    
    # Calculate Global Average for Relative Index
    global_avg = result_df['concentration'].mean()
    summary_df['relative_density_index'] = (summary_df['avg_density'] / global_avg).round(2)
    
    # Upload summary to SQL
    summary_df.to_sql('h2_proximity_analysis', con=engine, if_exists='replace', index=False)
    
    print("✅ Success! Results uploaded to 'h2_proximity_analysis'")
    print(summary_df.to_string(index=False))

except Exception as e:
    print(f"Error with sjoin_nearest: {e}")
    print("Falling back to a simplified SQL approach with a limit...")
    # Fallback: Just run the SQL query but limit the plastic data to 1000 rows to test
    # (Not recommended for final result, but good for debugging)

Loading data from SQL...
Converting to GeoDataFrames...
Calculating nearest neighbors (this may take a moment)...


c:\Users\Ready2Use\anaconda3\envs\river-risk-index\Lib\site-packages\geopandas\array.py:411: UserWarning: Geometry is in a geographic CRS. Results from 'sjoin_nearest' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  warnings.warn(


✅ Success! Results uploaded to 'h2_proximity_analysis'
    proximity_zone  sample_count  avg_density  max_hotspot  relative_density_index
Near River (<50km)         10774      49.1893      57665.0                     1.0


### Visualization Hypothesis 2

In [10]:
# 1. Load the result table
df = pd.read_sql("SELECT * FROM h2_proximity_analysis", engine)

# 2. Create a Bar Chart: Avg Density by Zone
fig_density = px.bar(
    df, 
    x='proximity_zone', 
    y='avg_density', 
    title='H2 Confirmation: Plastic Density Drops Sharply Away from Rivers',
    labels={'proximity_zone': 'Distance from River', 'avg_density': 'Avg Concentration (particles/m³)'},
    color='avg_density',
    color_continuous_scale='Reds'
)
fig_density.update_layout(showlegend=False)
fig_density.show()

# 3. Create a Pie Chart: Sample Distribution (Where is the data?)
fig_samples = px.pie(
    df, 
    values='sample_count', 
    names='proximity_zone', 
    title='H2 Confirmation: >99% of Samples Found Within 50km of Rivers',
    color_discrete_sequence=px.colors.sequential.Blues
)
fig_samples.show()

In [ ]:
# 1. Load Top 50 Rivers (The "Sources")

print("Loading Top 50 Emission Points...")
query_rivers = """
    SELECT lat, lon, emission_tons_year 
    FROM emission_points 
    ORDER BY emission_tons_year DESC 
    LIMIT 50
"""
rivers_df = pd.read_sql(query_rivers, engine)

print(f"Loaded {len(rivers_df)} top rivers.")

# 2. Load Plastic Data (The "Samples")
print("Loading Plastic Samples...")
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)
print(f"Loaded {len(plastic_df)} samples.")

# 3. Calculate Distances (Vectorized - SUPER FAST)
print("Calculating distances (this takes 1 second)...")
R = 6371.0 # Earth radius in km

# Convert to radians
rivers_rad = np.radians(rivers_df[['lat', 'lon']].values) # Ensure column name matches
plastic_rad = np.radians(plastic_df[['lat', 'lng']].values)

# Expand dimensions to calculate distance from EVERY plastic point to EVERY river point at once
# Shape: (15000, 1, 2) vs (1, 50, 2) -> Result (15000, 50)
dlat = plastic_rad[:, None, 0] - rivers_rad[None, :, 0]
dlon = plastic_rad[:, None, 1] - rivers_rad[None, :, 1]

a = np.sin(dlat/2.0)**2 + np.cos(plastic_rad[:, None, 0]) * np.cos(rivers_rad[None, :, 0]) * np.sin(dlon/2.0)**2
c = 2 * np.arcsin(np.sqrt(a))
distances_km = R * c

# Find the MINIMUM distance for each plastic point (nearest river)
min_distances = distances_km.min(axis=1)

plastic_df['dist_to_nearest_top_50_km'] = min_distances

# 4. Analyze
near_50km = plastic_df[plastic_df['dist_to_nearest_top_50_km'] <= 50]
total = len(plastic_df)
count_near = len(near_50km)
pct = (count_near / total) * 100

print(f"\n📊 H2 FINAL PROOF:")
print(f"   Total Samples: {total}")
print(f"   Samples within 50km of Top 50 Rivers: {count_near}")
print(f"   PERCENTAGE: {pct:.2f}%")

if pct > 50:
    print("   ✅ H2 CONFIRMED: Majority of coastal plastic is near the top rivers.")
else:
    print("   ⚠️ Result lower than expected. Maybe plastic is near smaller rivers (not top 50)?")

# 5. Visualize
fig = px.bar(
    x=['Within 50km of Top 50 Rivers', 'Farther Away'],
    y=[count_near, total - count_near],
    title=f'H2: {pct:.1f}% of Plastic Samples are Within 50km of Top 50 Rivers',
    color=[count_near, total - count_near],
    color_continuous_scale='Reds'
)
fig.show()

Loading Top 50 Emission Points...
Loaded 50 top rivers.
Loading Plastic Samples...
Loaded 10774 samples.
Calculating distances (this takes 1 second)...

📊 H2 FINAL PROOF:
   Total Samples: 10774
   Samples within 50km of Top 50 Rivers: 62
   PERCENTAGE: 0.58%
   ⚠️ Result lower than expected. Maybe plastic is near smaller rivers (not top 50)?


In [ ]:
# 1. Load 1,000 Rivers (Representing the "Long Tail")
print("Loading 1,000 River Emission Points...")
query_rivers = """
    SELECT lat, lon 
    FROM emission_points 
    ORDER BY emission_tons_year DESC 
    LIMIT 1000
"""

rivers_df = pd.read_sql(query_rivers, engine)

# 2. Load Plastic
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)

# 3. Calculate Distances (Same fast NumPy logic)
print("Calculating distances to nearest of 1,000 rivers...")
R = 6371.0
rivers_rad = np.radians(rivers_df[['lat', 'lon']].values if 'lon' in rivers_df.columns else rivers_df[['lat', 'lng']].values)
plastic_rad = np.radians(plastic_df[['lat', 'lng']].values)

dlat = plastic_rad[:, None, 0] - rivers_rad[None, :, 0]
dlon = plastic_rad[:, None, 1] - rivers_rad[None, :, 1]
a = np.sin(dlat/2.0)**2 + np.cos(plastic_rad[:, None, 0]) * np.cos(rivers_rad[None, :, 0]) * np.sin(dlon/2.0)**2
c = 2 * np.arcsin(np.sqrt(a))
distances_km = R * c
min_distances = distances_km.min(axis=1)

plastic_df['dist_to_nearest_river_km'] = min_distances

# 4. Analyze
near_50km = plastic_df[plastic_df['dist_to_nearest_river_km'] <= 50]
pct = (len(near_50km) / len(plastic_df)) * 100

print(f"\n📊 REVISED H2 PROOF (Using 1,000 rivers):")
print(f"   Samples within 50km of ANY of the Top 1,000 Rivers: {len(near_50km)}")
print(f"   PERCENTAGE: {pct:.2f}%")

# 5. Visualize
fig = px.bar(
    x=['Within 50km of Top 1,000 Rivers', 'Farther Away'],
    y=[len(near_50km), len(plastic_df) - len(near_50km)],
    title=f'Revised H2: {pct:.1f}% of Plastic is Near River Mouths (Not just Top 50)',
    color=[len(near_50km), len(plastic_df) - len(near_50km)],
    color_continuous_scale='Greens'
)
fig.show()

Loading 1,000 River Emission Points...
Calculating distances to nearest of 1,000 rivers...

📊 REVISED H2 PROOF (Using 1,000 rivers):
   Samples within 50km of ANY of the Top 1,000 Rivers: 325
   PERCENTAGE: 3.02%


### Hypothesis 3

In [13]:
# 1. Load Config
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)

# 2. Load Parquet (File 8)
file_path = config["output_data"]["file8"]
print(f"Loading: {file_path}")
cleanup_df = pd.read_parquet(file_path)

# 3. Inspect Columns
print("Columns:", cleanup_df.columns.tolist())
print(cleanup_df.head())

# 4. Upload to SQL
# Ensure column names are SQL-friendly
cleanup_df.columns = [c.lower().replace(' ', '_') for c in cleanup_df.columns]

cleanup_df.to_sql('ocean_cleanup_efforts', con=engine, if_exists='replace', index=False)
print(f"✅ Uploaded {len(cleanup_df)} rows to 'ocean_cleanup_efforts'")

Loading: ../Data/Clean/ocean_cleanup_efforts.parquet
Columns: ['year', 'organisation', 'cleanup_type', 'source_url', 'kg_removed_annual', 'kg_removed_cumulative', 'volunteers', 'countries']
   year             organisation      cleanup_type  \
0  2008  Ocean Conservancy (ICC)  Beach + Waterway   
1  2009  Ocean Conservancy (ICC)  Beach + Waterway   
2  2010  Ocean Conservancy (ICC)  Beach + Waterway   
3  2011  Ocean Conservancy (ICC)  Beach + Waterway   
4  2012  Ocean Conservancy (ICC)  Beach + Waterway   

                                          source_url  kg_removed_annual  \
0  https://oceanconservancy.org/work/plastics/cle...            3265862   
1  https://oceanconservancy.org/work/plastics/cle...            3447299   
2  https://oceanconservancy.org/work/plastics/cle...            3628736   
3  https://oceanconservancy.org/work/plastics/cle...            4173046   
4  https://oceanconservancy.org/work/plastics/cle...            4626638   

   kg_removed_cumulative  voluntee

### Map 5 Gyres

In [ ]:
# 1. Load ALL data to see the global distribution
print("Loading all plastic data...")
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)

# Check distribution
print(f"Total points: {len(plastic_df)}")
print("Points by Latitude Zone:")
plastic_df['Zone'] = pd.cut(plastic_df['lat'], bins=[-90, -30, 30, 90], labels=['Southern Ocean', 'Tropics', 'Northern Temperate'])
print(plastic_df['Zone'].value_counts())

# 2. Define Gyre Centers
gyres_df = pd.DataFrame({
    'Gyre': ['North Pacific', 'South Pacific', 'North Atlantic', 'South Atlantic', 'Indian Ocean'],
    'lat': [40, -35, 40, -30, -25],
    'lng': [-160, -110, -40, -15, 80]
})

# 3. Create Map
fig = px.scatter_mapbox(
    plastic_df,
    lat="lat",
    lon="lng",
    color="concentration",
    color_continuous_scale="Reds",
    size_max=10, # Bigger dots
    opacity=0.8,
    title="Q2: Global Plastic Distribution (Color Scale Clipped to 1-6)",
    mapbox_style="carto-positron",
    zoom=1,
    center={"lat": 20, "lon": 0}
)
# If the max is 57,000, everything < 10 looks white. This forces the scale to show 0-5.
fig.update_coloraxes(cmax=5, cmin=0)

# Add Gyres as Blue Stars
fig.add_trace(px.scatter_mapbox(
    gyres_df, lat="lat", lon="lng",
    color_discrete_sequence=["blue"],
    size_max=15
).data[0])

# Add Labels
for i, row in gyres_df.iterrows():
    fig.add_annotation(x=row['lng'], y=row['lat'], text=row['Gyre'],
                       font=dict(color="blue", size=12, family="Arial Black"),
                       showarrow=False, bgcolor="white", bordercolor="blue")

fig.update_layout(margin={"r":0,"t":50,"l":0,"b":0})
fig.show()

Loading all plastic data...
Total points: 10774
Points by Latitude Zone:
Zone
Tropics               5568
Northern Temperate    4825
Southern Ocean         381
Name: count, dtype: int64


C:\Users\Ready2Use\AppData\Local\Temp\ipykernel_22908\2939872565.py:19: DeprecationWarning: *scatter_mapbox* is deprecated! Use *scatter_map* instead. Learn more at: https://plotly.com/python/mapbox-to-maplibre/
  fig = px.scatter_mapbox(
C:\Users\Ready2Use\AppData\Local\Temp\ipykernel_22908\2939872565.py:37: DeprecationWarning: *scatter_mapbox* is deprecated! Use *scatter_map* instead. Learn more at: https://plotly.com/python/mapbox-to-maplibre/
  fig.add_trace(px.scatter_mapbox(


here we can see where the plastic accumulates
1 north pacific
2 south pacific
3 north atlantic
4 south atlantic
5 indian ocean

### Bivariate analysis: source volume vs accumulation zone.

In [ ]:
# 1. Load Data
print("Loading Emission Points (Top 1000)...")
emission_df = pd.read_sql("""
    SELECT lat, lon, emission_tons_year 
    FROM emission_points 
    ORDER BY emission_tons_year DESC 
    LIMIT 1000
""", engine)

print("Loading Plastic Samples...")
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)

# 2. Calculate Distance to Nearest River (Vectorized)
print("Calculating distances (this takes a moment)...")
R = 6371.0
rivers_rad = np.radians(emission_df[['lat', 'lon']].values)
plastic_rad = np.radians(plastic_df[['lat', 'lng']].values)

dlat = plastic_rad[:, None, 0] - rivers_rad[None, :, 0]
dlon = plastic_rad[:, None, 1] - rivers_rad[None, :, 1]
a = np.sin(dlat/2.0)**2 + np.cos(plastic_rad[:, None, 0]) * np.cos(rivers_rad[None, :, 0]) * np.sin(dlon/2.0)**2
c = 2 * np.arcsin(np.sqrt(a))
distances = R * c
min_dist = distances.min(axis=1)

plastic_df['dist_to_nearest_major_river'] = min_dist

# 3. Filter: Only plastic within 50km (Coastal Accumulation)
threshold = 50
coastal_plastic = plastic_df[plastic_df['dist_to_nearest_major_river'] <= threshold].copy()

if len(coastal_plastic) == 0:
    print(f"⚠️ No samples found within {threshold}km. Expanding to 200km...")
    threshold = 200
    coastal_plastic = plastic_df[plastic_df['dist_to_nearest_major_river'] <= threshold].copy()

print(f"Analyzing {len(coastal_plastic)} samples within {threshold}km of top 1000 rivers.")

# 4. Link Plastic to Nearest River's Emission Volume
min_indices = distances.argmin(axis=1)
# Map indices only for the filtered dataframe
filtered_indices = plastic_df.index.isin(coastal_plastic.index)
# We need to re-calculate min_indices for the filtered set to match the emission_df
# Actually, we can just use the global min_indices and filter them
coastal_plastic['nearest_river_idx'] = min_indices[filtered_indices]

emission_volumes = emission_df['emission_tons_year'].values
coastal_plastic['source_emission'] = emission_volumes[coastal_plastic['nearest_river_idx']]

# 5. Bin Emissions
coastal_plastic['emission_bin'] = pd.qcut(coastal_plastic['source_emission'], q=4, labels=['Low (Q1)', 'Medium (Q2)', 'High (Q3)', 'Very High (Q4)'])

# 6. Aggregate
summary = coastal_plastic.groupby('emission_bin', observed=True).agg(
    sample_count=('concentration', 'count'),
    avg_density=('concentration', 'mean'),
    max_density=('concentration', 'max'),
    avg_emission=('source_emission', 'mean')
).reset_index()

print("\n📊 Bivariate Analysis Summary:")
print(summary.to_string(index=False))

# 7. Visualization 1: Bar Chart
fig = px.bar(
    summary,
    x='emission_bin',
    y='avg_density',
    title=f'River Emission Volume vs. Coastal Plastic Density\n(Avg Density within {threshold}km)',
    labels={'emission_bin': 'River Emission Category', 'avg_density': 'Avg Density (particles/m³)'},
    color='avg_density',
    color_continuous_scale='Reds',
    text_auto='.2f'
)
fig.update_layout(showlegend=False)
fig.show()

# 8. Visualization 2: Scatter Plot 
if len(coastal_plastic) < 5000:
    fig2 = px.scatter(
        coastal_plastic,
        x='source_emission',
        y='concentration',
        title='Individual Samples: Source Emission vs. Observed Density',
        labels={'source_emission': 'River Emission (Tons/Year)', 'concentration': 'Plastic Density'},
        opacity=0.3,
        # trendline='ols'  <-- REMOVED THIS LINE
        log_x=True # Use log scale for X because emission ranges are huge
    )
    fig2.update_traces(marker=dict(size=8))
    fig2.show()
else:
    print("Skipping scatter plot (too many points).")

Loading Emission Points (Top 1000)...
Loading Plastic Samples...
Calculating distances (this takes a moment)...
Analyzing 325 samples within 50km of top 1000 rivers.

📊 Bivariate Analysis Summary:
  emission_bin  sample_count  avg_density  max_density  avg_emission
      Low (Q1)           104  1060.076910 10133.333333    235.782596
   Medium (Q2)            92     0.522118     6.800000    318.230837
     High (Q3)            57     1.180980    21.300000    509.994737
Very High (Q4)            72  1148.542209 57665.000000   3370.250514


The analysis reveals a U-shaped correlation. Both the smallest and largest rivers create high-density coastal hotspots, while medium-sized rivers show low local density. This suggests that geography and hydrodynamics (trapping vs. flushing) are just as critical as emission volume. Small rivers in trapped bays are surprisingly dangerous accumulation zones.

Volume ≠ Destiny: A river with 100 tons/year emission can have 10,000 density (if it's in a trapped bay), while a river with 1,000 tons/year might have 0 density (if it's on a windy coast).
The "Black Swan" Events: The few dots at 50k-60k density represent extreme pollution events (e.g., a specific harbor or river mouth that is completely choked). These are the critical targets for cleanup.

The bivariate analysis disproves a simple linear relationship. High-density hotspots exist at both ends of the spectrum: small rivers in enclosed bays and massive rivers with high throughput. Therefore, a successful global strategy must target both the 'Top 50' mega-rivers and the 'High-Risk Bays' fed by smaller rivers.

In [22]:
# 1. Load data
emission_df = pd.read_sql("SELECT lat, lon, emission_tons_year FROM emission_points ORDER BY emission_tons_year DESC LIMIT 1000", engine)
plastic_df = pd.read_sql("SELECT lat, lng, concentration FROM observed_plastic WHERE concentration > 0", engine)

# 2. Calculate Distances (Vectorized)
R = 6371.0
rivers_rad = np.radians(emission_df[['lat', 'lon' if 'lon' in emission_df.columns else 'lng']].values)
plastic_rad = np.radians(plastic_df[['lat', 'lng']].values)

dlat = plastic_rad[:, None, 0] - rivers_rad[None, :, 0]
dlon = plastic_rad[:, None, 1] - rivers_rad[None, :, 1]
a = np.sin(dlat/2.0)**2 + np.cos(plastic_rad[:, None, 0]) * np.cos(rivers_rad[None, :, 0]) * np.sin(dlon/2.0)**2
c = 2 * np.arcsin(np.sqrt(a))
distances = R * c
min_dist = distances.min(axis=1)

plastic_df['dist'] = min_dist

# 3. Filter Coastal (50km)
coastal = plastic_df[plastic_df['dist'] <= 50].copy()
if len(coastal) == 0:
    print("⚠️ No data within 50km. Expanding to 200km...")
    coastal = plastic_df[plastic_df['dist'] <= 200].copy()

# 4. Link to Emission Volume
min_indices = distances.argmin(axis=1)
# Safe indexing
valid_indices = plastic_df.index.isin(coastal.index)
coastal['river_idx'] = min_indices[valid_indices]
coastal['emission'] = emission_df['emission_tons_year'].values[coastal['river_idx']]

# 5. Create Bins
# duplicates='drop' handles cases where qcut can't split evenly
coastal['bin'] = pd.qcut(coastal['emission'], q=4, labels=['Q1', 'Q2', 'Q3', 'Q4'], duplicates='drop')

# 6. Clean Data
coastal = coastal.dropna(subset=['emission', 'bin', 'concentration'])

print(f"📊 Analyzing {len(coastal)} coastal samples...")
print(f"   Groups found: {coastal['bin'].unique()}")

if len(coastal['bin'].unique()) < 2:
    print("❌ Error: Not enough groups to compare. Cannot run ANOVA.")
else:
    # --- TEST 1: Pearson ---
    try:
        pearson_corr, p_val_pearson = stats.pearsonr(coastal['emission'], coastal['concentration'])
        print(f"\n1. PEARSON CORRELATION (Linear)")
        print(f"   r = {pearson_corr:.4f}, p = {p_val_pearson:.4e}")
    except Exception as e:
        print(f"\n1. PEARSON: Could not calculate ({e})")
        pearson_corr, p_val_pearson = 0, 1.0

    # --- TEST 2: Spearman ---
    try:
        spearman_corr, p_val_spearman = stats.spearmanr(coastal['emission'], coastal['concentration'])
        print(f"\n2. SPEARMAN CORRELATION (Rank)")
        print(f"   rho = {spearman_corr:.4f}, p = {p_val_spearman:.4e}")
    except Exception as e:
        print(f"\n2. SPEARMAN: Could not calculate ({e})")
        spearman_corr, p_val_spearman = 0, 1.0

    # --- TEST 3: ANOVA ---
    try:
        groups = [group['concentration'].values for name, group in coastal.groupby('bin')]
        f_stat, p_val_anova = stats.f_oneway(*groups)
        print(f"\n3. ONE-WAY ANOVA (Group Differences)")
        print(f"   F = {f_stat:.4f}, p = {p_val_anova:.4e}")
        
        # --- FINAL VERDICT ---
        print("\n" + "="*50)
        print("📝 STATISTICAL CONCLUSION")
        print("="*50)
        if p_val_anova < 0.05:
            print("✅ SIGNIFICANT DIFFERENCE FOUND between river categories.")
            print("   (The U-Shape in your bar chart is statistically real.)")
        else:
            print("⚠️ No significant difference between groups.")
            
        if p_val_pearson > 0.05:
            print("✅ NO LINEAR CORRELATION found.")
            print("   (Confirms that 'Bigger River' does not equal 'Higher Density' in a straight line.)")
            
    except Exception as e:
        print(f"\n3. ANOVA: Could not calculate ({e})")

📊 Analyzing 325 coastal samples...
   Groups found: ['Q3', 'Q2', 'Q4', 'Q1']
Categories (4, object): ['Q1' < 'Q2' < 'Q3' < 'Q4']

1. PEARSON CORRELATION (Linear)
   r = 0.2321, p = 2.3701e-05

2. SPEARMAN CORRELATION (Rank)
   rho = -0.0111, p = 8.4174e-01

3. ONE-WAY ANOVA (Group Differences)
   F = 2.7727, p = 4.1587e-02

📝 STATISTICAL CONCLUSION
✅ SIGNIFICANT DIFFERENCE FOUND between river categories.
   (The U-Shape in your bar chart is statistically real.)


C:\Users\Ready2Use\AppData\Local\Temp\ipykernel_22908\206603176.py:65: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  groups = [group['concentration'].values for name, group in coastal.groupby('bin')]


Why only 325 samples: 

1. The Filter: "Coastal Proximity"
We applied a strict filter:

"Only analyze plastic samples found within 50km of one of the Top 1,000 Rivers."

Total Samples: 10,774
Samples > 50km from a major river: ~10,449 (These are in the open ocean, deep coastal zones, or near tiny creeks not in the Top 1,000).
Samples ≤ 50km from a Top 1,000 River: 325.
Why so few?

Ocean Scale: 50km is a tiny distance in the ocean. Most of your "Coastal" data is likely 100km+ off the shore or along coastlines that don't have a "Top 1,000" river nearby.
Dataset Bias: As we saw in the map, your data is clustered in specific regions (e.g., US East Coast, Mediterranean). Many of these clusters might be near small rivers (not in the Top 1,000) or far from any major river mouth.
The "Top 1,000" Limit: We only looked at the top 1,000 emitters. If a sample is near the 1,001st river, it gets excluded.
2. Is 325 Enough? (Statistical Power)
Yes.

In statistics, quality > quantity.
We are comparing 4 groups (Q1, Q2, Q3, Q4).
With 325 samples, you have roughly 81 samples per group.
Rule of Thumb: You generally need ~30 samples per group to run a valid ANOVA. You have nearly triple that.
The Proof: The ANOVA p-value was 0.042 (significant). If the sample size were too small, the test would lack the power to find a difference, and the p-value would be > 0.05. The fact that we found significance proves the sample size was sufficient.

To validate the relationship between river emission volume and coastal plastic density, three statistical tests were performed on 325 coastal samples:

1. Pearson Correlation (r=0.23,p<0.001)
What it tests: "Is there a straight line where density goes up as emission goes up?"
The Result: $0.23 is a ∗∗weak positive∗∗ correlation. The p−value(0.00002) says this weak trend is real, not random.
Translation: There is a slight tendency for bigger rivers to have higher density, BUT the relationship is very messy. 
If you tried to draw a straight line through your scatter plot, it would be a terrible fit. 
This confirms your visual finding that volume alone isn't the whole story.

2. Spearman Rank Correlation (ρ=−0.01,p=0.84)
What it tests: "As river size increases, does the rank of pollution density consistently increase?" 
(This ignores the exact numbers and just looks at the order).
The Result: −0.01 is effectively ZERO. The p-value (0.84) is huge, meaning no statistical significance.
Translation: There is no consistent monotonic trend.
This is the "smoking gun" for your U-Shape.
If the relationship were linear (Small→Med→High→Very High), Spearman would be high (e.g., 0.8).
Because your data goes High (Q1) → Low (Q2/Q3) → High (Q4), the "ranks" cancel each other out. 
The small rivers (Q1) are high, which breaks the pattern that "bigger is always worse."
Conclusion: You cannot predict pollution density just by ranking rivers by size.

3. One-Way ANOVA (F=2.77,p=0.042)
What it tests: "Are the average densities of the four groups (Q1, Q2, Q3, Q4) significantly different from each other?"
The Result: The p-value is 0.042. Since this is < 0.05, the result is Statistically Significant.
Translation: The differences you see in your bar chart (1060 vs 0.5 vs 1148) are NOT due to random chance.
It proves that River Category matters.
Specifically, it validates that the Q1 (Small) and Q4 (Very Large) groups are genuinely distinct from the Q2/Q3 (Medium) groups.
Without this test, a critic could say "Maybe those high bars are just luck." This test says: "No, the difference is real."

Conclusion: The data rejects the hypothesis that plastic density scales linearly with river emission volume. 
Instead, it supports a bimodal accumulation model, where both small rivers in enclosed geographies and massive 
high-throughput rivers create significant coastal hotspots, while medium-sized rivers contribute less to local coastal density.

In [ ]:
# 1. open file
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)

# Get file path from config
file_path = config["output_data"]["file5"]
# 2. Load Parquet Data
ocean = pd.read_parquet(file_path)

print(ocean.shape)
print(ocean.dtypes)
print(ocean.head(3))

# 1. open file
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)

# Get file path from config
file_path = config["output_data"]["file1"]
# 2. Load Parquet Data
rivers = pd.read_parquet(file_path)

print(rivers.shape)
print(rivers.dtypes)
print(rivers.head(3))

(80, 6)
country                          object
iso_code                         object
year                              Int64
plastic_ocean_tonnes            float64
plastic_ocean_million_tonnes    float64
decade                            int64
dtype: object
  country iso_code  year  plastic_ocean_tonnes  plastic_ocean_million_tonnes  \
0   China      CHN  2000              683704.0                        0.6837   
1   China      CHN  2001              767961.0                        0.7680   
2   China      CHN  2002              860436.0                        0.8604   

   decade  
0    2000  
1    2000  
2    2000  
(31819, 4)
emission     float64
geometry      object
country       object
continent     object
dtype: object
   emission                                           geometry      country  \
0  0.164904  b'\x01\x01\x00\x00\x008\x87\x88\x88\x88\x19e@ ...  New Zealand   
1  0.124932  b'\x01\x01\x00\x00\x00\x0c\x8e\xc2\xf5(\x0be@p...  New Zealand   
2  1.213370  b'\x01\x01

In [18]:
# ============================================================
# Bivariate Analysis: River Emission Source vs Ocean Plastic Accumulation
# Q2 — Where does plastic accumulate?
# ============================================================

import pandas as pd
import geopandas as gpd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from scipy import stats
from shapely import wkb



# ── 2. DESERIALIZE GEOMETRY ──────────────────────────────────

rivers['geometry'] = rivers['geometry'].apply(lambda b: wkb.loads(b))
gdf = gpd.GeoDataFrame(rivers, geometry='geometry', crs='EPSG:4326')

# ── 3. AGGREGATE RIVERS TO COUNTRY LEVEL ─────────────────────

river_by_country = (
    gdf.groupby('country')
    .agg(
        total_emission       = ('emission', 'sum'),
        emission_point_count = ('emission', 'count'),
        mean_emission        = ('emission', 'mean'),
        continent            = ('continent', 'first')
    )
    .reset_index()
)

# ── 4. AGGREGATE OCEAN PLASTIC TO COUNTRY LEVEL ──────────────
# Use most recent decade for a stable signal (avoids single noisy year)

ocean_agg = (
    ocean[ocean['decade'] == ocean['decade'].max()]
    .groupby(['country', 'iso_code'])
    .agg(
        avg_ocean_plastic_tonnes = ('plastic_ocean_tonnes', 'mean')
    )
    .reset_index()
)

# ── 5. JOIN ───────────────────────────────────────────────────

merged = river_by_country.merge(ocean_agg, on='country', how='inner')

print(f"Matched countries: {len(merged)}")
print(f"River-only countries (no ocean data): {len(river_by_country) - len(merged)}")
print(merged[['country', 'total_emission', 'avg_ocean_plastic_tonnes']].head(10))

# ── 6. CORRELATION ANALYSIS ───────────────────────────────────

# Log-transform both — emission data is heavily right-skewed
merged['log_emission']     = np.log1p(merged['total_emission'])
merged['log_ocean_plastic'] = np.log1p(merged['avg_ocean_plastic_tonnes'])

pearson_r, pearson_p   = stats.pearsonr(merged['log_emission'], merged['log_ocean_plastic'])
spearman_r, spearman_p = stats.spearmanr(merged['total_emission'], merged['avg_ocean_plastic_tonnes'])

print("\n── Correlation Results ──────────────────────────────")
print(f"Pearson  r (log-log): {pearson_r:.3f}  p={pearson_p:.4f}")
print(f"Spearman r (raw):     {spearman_r:.3f}  p={spearman_p:.4f}")
print("────────────────────────────────────────────────────")

# ── 7. SCATTER PLOT WITH REGRESSION ───────────────────────────

fig_scatter = px.scatter(
    merged,
    x='log_emission',
    y='log_ocean_plastic',
    color='continent',
    hover_name='country',
    hover_data={
        'total_emission': ':.2f',
        'avg_ocean_plastic_tonnes': ':.0f',
        'log_emission': False,
        'log_ocean_plastic': False
    },
    trendline='ols',
    trendline_scope='overall',
    labels={
        'log_emission':      'River emission (log tonnes/year)',
        'log_ocean_plastic': 'Ocean plastic (log tonnes, latest decade avg)',
        'continent':         'Continent'
    },
    title=f'River emissions vs ocean plastic accumulation by country<br>'
          f'<sup>Pearson r={pearson_r:.3f} (log-log) | Spearman r={spearman_r:.3f} | n={len(merged)}</sup>'
)

fig_scatter.update_traces(marker=dict(size=8, opacity=0.75))
fig_scatter.update_layout(height=550)
fig_scatter.show()

# ── 8. CHOROPLETH MAP — RIVER EMISSIONS ───────────────────────

fig_map_emission = px.choropleth(
    merged,
    locations='iso_code',
    color='total_emission',
    hover_name='country',
    color_continuous_scale='YlOrRd',
    title='Total river plastic emissions by country (Meijer 2021)',
    labels={'total_emission': 'Emission (tonnes/year)'}
)
fig_map_emission.update_layout(height=450)
fig_map_emission.show()

# ── 9. CHOROPLETH MAP — OCEAN PLASTIC ACCUMULATION ────────────

fig_map_ocean = px.choropleth(
    merged,
    locations='iso_code',
    color='avg_ocean_plastic_tonnes',
    hover_name='country',
    color_continuous_scale='Blues',
    title='Ocean plastic accumulation by country (latest decade avg)',
    labels={'avg_ocean_plastic_tonnes': 'Ocean plastic (tonnes)'}
)
fig_map_ocean.update_layout(height=450)
fig_map_ocean.show()

# ── 10. BUBBLE MAP — BOTH DIMENSIONS TOGETHER ─────────────────
# Emission = colour intensity, ocean plastic = bubble size

fig_bubble = px.scatter_geo(
    merged,
    locations='iso_code',
    size='avg_ocean_plastic_tonnes',
    color='total_emission',
    hover_name='country',
    color_continuous_scale='YlOrRd',
    size_max=40,
    projection='natural earth',
    title='Source volume (colour) vs accumulation (bubble size) by country',
    labels={
        'total_emission':            'River emission (tonnes/year)',
        'avg_ocean_plastic_tonnes':  'Ocean plastic (tonnes)'
    }
)
fig_bubble.update_layout(height=500)
fig_bubble.show()

# ── 11. SUMMARY TABLE ─────────────────────────────────────────

summary = (
    merged[['country', 'continent', 'total_emission',
            'avg_ocean_plastic_tonnes', 'emission_point_count']]
    .sort_values('total_emission', ascending=False)
    .head(20)
    .reset_index(drop=True)
)
summary.index += 1
print("\nTop 20 countries by river emission vs their ocean plastic load:")
print(summary.to_string())

Matched countries: 2
River-only countries (no ocean data): 136
  country  total_emission  avg_ocean_plastic_tonnes
0   China    70762.513282               2864860.800
1   India   126476.891778                838552.791

── Correlation Results ──────────────────────────────
Pearson  r (log-log): -1.000  p=1.0000
Spearman r (raw):     -1.000  p=nan
────────────────────────────────────────────────────



Top 20 countries by river emission vs their ocean plastic load:
  country continent  total_emission  avg_ocean_plastic_tonnes  emission_point_count
1   India      Asia   126476.891778                838552.791                  1166
2   China      Asia    70762.513282               2864860.800                  1309


In [30]:
# 1. Load Configuration and Data
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)

file_path = config["output_data"]["file3"] # marine_microplastics.parquet
df = pd.read_parquet(file_path)

display(df.head())


,OBJECTID,lat,lng,ocean,setting,sampling_method,mesh_size_mm,microplastics_measurement,unit,concentration_class,sample_date,x,y,sample_datetime
0,1,45.28,-60.29,Atlantic Ocean,Ocean water,Neuston net,0.335,0.000000,pieces/m3,Very Low,1989-07-13,-60.29,45.28,1989-07-13
1,2,40.93,-70.65,Atlantic Ocean,Ocean water,Neuston net,0.335,0.002276,pieces/m3,Low,1989-10-12,-70.65,40.93,1989-10-12
2,3,40.93,-70.65,Atlantic Ocean,Ocean water,Neuston net,0.335,0.004320,pieces/m3,Low,1989-10-12,-70.65,40.93,1989-10-12
3,4,40.30,-69.77,Atlantic Ocean,Ocean water,Neuston net,0.335,0.000000,pieces/m3,Very Low,1989-10-12,-69.77,40.30,1989-10-12
4,5,39.88,-67.15,Atlantic Ocean,Ocean water,Neuston net,0.335,0.000000,pieces/m3,Very Low,1989-10-15,-67.15,39.88,1989-10-15


In [32]:
# 2. Extract Year from 'sample_date'
print("Extracting year from 'sample_date'...")
df['sample_date'] = pd.to_datetime(df['sample_date'], errors='coerce')
df['year'] = df['sample_date'].dt.year

# Ensure concentration is numeric
df['microplastics_measurement'] = pd.to_numeric(df['microplastics_measurement'], errors='coerce')

# Drop rows where year or concentration is missing
df = df.dropna(subset=['year', 'microplastics_measurement'])
df['year'] = df['year'].astype(int)


Extracting year from 'sample_date'...


C:\Users\Ready2Use\AppData\Local\Temp\ipykernel_22908\2482235136.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['year'] = df['year'].astype(int)


In [ ]:
print("📅 Year Range Check:")
print(f"   Earliest Year: {df['year'].min()}")
print(f"   Latest Year:   {df['year'].max()}")

print("\n📊 Sample Count per Year:")
yearly_counts = df.groupby('year').size()
print(yearly_counts)

# Visualize the sample volume per year to see the "2013 surge"

fig_vol = px.bar(
    x=yearly_counts.index, 
    y=yearly_counts.values,
    title='Number of Samples per Year (Total Dataset)',
    labels={'x': 'Year', 'y': 'Sample Count'},
    color=yearly_counts.values,
    color_continuous_scale='reds'
)
fig_vol.show()

📅 Year Range Check:
   Earliest Year: 1972
   Latest Year:   2023

📊 Sample Count per Year:
year
1972      31
1973      14
1986      20
1987     161
1989     180
1990     110
1991     294
1992     517
1993     362
1994     324
1995     350
1996     408
1997     388
1998     415
1999     355
2000     445
2001     474
2002     446
2003     419
2004     401
2005     492
2006     339
2007     443
2008     440
2009     420
2010     315
2011     570
2012     619
2013     370
2014    1217
2015    1114
2016     621
2017     737
2018    1097
2019     822
2020     455
2021     451
2022      52
2023      25
dtype: int64


In [37]:
# 3. Inspect Sampling Methods
print("Available Sampling Methods:")
print(df['sampling_method'].unique())
print(f"Total Rows: {len(df)}")

# 4. Define Groups based on your observation

# Common Net methods: 'Neuston net', 'Manta net', 'Plankton net'
# Common Container methods: 'Jar', 'Bottle', 'Spoon', 'Grab sample'
net_keywords = [
    'Neuston net', 'Manta net', 'AVANI net', 'Plankton net', 
    'plankton net', 'manta net', 'net' 
]

container_keywords = [
    'Grab sample', 'Intake seawater pump', 'Stainless steel bucket', 'Aluminum bucket', 
    'CTD rosette sampler', 'Day grab', 'Ekman dredge', 'Megacorer', 'Stainless steel spoon', 
    'Trowel', 'Metal scoop', 'Metal spoon', 'stainless-steel sampler', 'Petite Ponar benthic grab', 
    'Sediment grab sampler', 'Van Veen grab', 'shovel', 'Van Dorn sampler', 'Shipek grab sampler',
    'glass jar', 'Kubiena box', 'stainless-steel spatula', 'stainless steel spatula', 'stainless steel bucket',
    'Niskin bottle', 'box corer', 'Ekman grab', 'PVC cylinder'
]

# Filter Data
df_net = df[df['sampling_method'].astype(str).str.contains('|'.join(net_keywords), na=False)]
df_container = df[df['sampling_method'].astype(str).str.contains('|'.join(container_keywords), na=False)]

print(f"\n Data Split:")
print(f"   Net Samples: {len(df_net)}")
print(f"   Container Samples: {len(df_container)}")

# 5. Aggregate by Year for both groups
def aggregate_by_year(data, group_name):
    if len(data) == 0:
        return pd.DataFrame()
    agg = data.groupby('year')['microplastics_measurement'].mean().reset_index()
    agg.columns = ['year', 'avg_microplastics_measurement']
    agg['group'] = group_name
    return agg

stats_net = aggregate_by_year(df_net, 'Net Sampling (Standard)')
stats_container = aggregate_by_year(df_container, 'Container Sampling (Micro/High-Res)')

# Combine for plotting
combined_stats = pd.concat([stats_net, stats_container], ignore_index=True)

# 6. Create the "Contrast" Chart
fig = go.Figure()

# Plot Net Data (The "Stable" Truth)
fig.add_trace(go.Scatter(
    x=stats_net['year'], 
    y=stats_net['avg_microplastics_measurement'], 
    mode='lines+markers', 
    name='Net Sampling (Standard)',
    line=dict(color='green', width=3),
    marker=dict(size=8)
))

# Plot Container Data (The "Shock" Spike)
fig.add_trace(go.Scatter(
    x=stats_container['year'], 
    y=stats_container['avg_microplastics_measurement'], 
    mode='lines+markers', 
    name='Container/Jar Sampling (High-Res)',
    line=dict(color='red', width=3, dash='dot'),
    marker=dict(size=8)
))

fig.update_layout(
    title='<b>Methodological Bias: Why Raw Trends Are Misleading</b><br><sup>Comparing Net vs. Container Sampling Methods</sup>',
    xaxis_title='Year',
    yaxis_title='Avg microplastics_measurement (particles/m³)',
    yaxis_type='log', # Log scale is crucial here because the difference might be 1000x
    hovermode='x unified',
    template='plotly_white',
    height=600
)

fig.show()

# 7. Print the "Shock" Stats for your report
if not stats_container.empty and not stats_net.empty:
    max_net = stats_net['avg_microplastics_measurement'].max()
    max_container = stats_container['avg_microplastics_measurement'].max()
    ratio = max_container / max_net if max_net > 0 else 0
    
    print(f"\n Portfolio Insight:")
    print(f"   Max Avg (Net): {max_net:.4f}")
    print(f"   Max Avg (Container): {max_container:.4f}")
    print(f"   The 'Shock' Factor: Container methods show ~{ratio:.1f}x higher density!")
    print("    Conclusion: Container methods provide a more accurate representation of the total plastic burden (including microplastics), " \
    "suggesting that historical net-based data significantly underestimated the true scale of ocean pollution.")

Available Sampling Methods:
['Neuston net' 'Grab sample' 'Intake seawater pump' 'Manta net'
 'Surface water intake' 'AVANI net' 'Plankton net'
 'Stainless steel bucket' 'Aluminum bucket' 'CTD rosette sampler'
 'Remotely operated vehicle' 'Day grab' 'Ekman dredge' 'Megacorer'
 'Stainless steel spoon' 'Trowel' 'Metal scoop' 'Metal spoon'
 'PVC cylinder' 'stainless-steel sampler' 'Petite Ponar benthic grab'
 'Sediment grab sampler' 'Hand picking' 'Van Veen grab' 'plankton net'
 'shovel' 'Van Dorn sampler' 'Shipek grab sampler' 'glass jar'
 'Kubiena box' 'stainless-steel spatula' ' stainless steel spatula'
 'manta net' 'stainless steel bucket' 'Niskin bottle' 'box corer'
 'Ekman grab']
Total Rows: 16713

 Data Split:
   Net Samples: 13385
   Container Samples: 3265



 Portfolio Insight:
   Max Avg (Net): 2.5546
   Max Avg (Container): 44352.0000
   The 'Shock' Factor: Container methods show ~17361.9x higher density!
    Conclusion: Container methods provide a more accurate representation of the total plastic burden (including microplastics), suggesting that historical net-based data significantly underestimated the true scale of ocean pollution.


In [38]:
# 8. Calculate the "Shift" in Methodology
df['method_type'] = 'Other'
df.loc[df['sampling_method'].astype(str).str.contains('|'.join(net_keywords), case=False), 'method_type'] = 'Net'
df.loc[df['sampling_method'].astype(str).str.contains('|'.join(container_keywords), case=False), 'method_type'] = 'Container'

# Group by Year and Method
method_trend = df.groupby(['year', 'method_type']).size().unstack(fill_value=0)
method_trend['Container_Pct'] = (method_trend['Container'] / (method_trend['Net'] + method_trend['Container'] + 1e-9)) * 100

print("\n📈 Methodology Shift Over Time:")
print(method_trend[['Net', 'Container', 'Container_Pct']].head(10)) # Show first 10 years
print("...")
print(method_trend[['Net', 'Container', 'Container_Pct']].tail(5)) # Show last 5 years

# Optional: Plot the % shift
fig2 = go.Figure()
fig2.add_trace(go.Scatter(
    x=method_trend.index, 
    y=method_trend['Container_Pct'], 
    mode='lines+markers', 
    name='% Container Samples',
    line=dict(color='purple', width=3),
    fill='tozeroy'
))

fig2.update_layout(
    title='<b>The Shift in Sampling Methods (2006–Present)</b><br><sup>Rise of Container Sampling explains the Density Spike</sup>',
    xaxis_title='Year',
    yaxis_title='% of Samples using Containers',
    template='plotly_white',
    height=400
)
fig2.show()

C:\Users\Ready2Use\AppData\Local\Temp\ipykernel_22908\4185678758.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['method_type'] = 'Other'



📈 Methodology Shift Over Time:
method_type  Net  Container  Container_Pct
year                                      
1972          31          0            0.0
1973          14          0            0.0
1986          20          0            0.0
1987         161          0            0.0
1989         180          0            0.0
1990         110          0            0.0
1991         294          0            0.0
1992         517          0            0.0
1993         362          0            0.0
1994         324          0            0.0
...
method_type  Net  Container  Container_Pct
year                                      
2019         526        253      32.477535
2020         288        167      36.703297
2021         357         94      20.842572
2022          52          0       0.000000
2023           0         25     100.000000


In [39]:
print(sorted(df['year'].unique()))

[np.int64(1972), np.int64(1973), np.int64(1986), np.int64(1987), np.int64(1989), np.int64(1990), np.int64(1991), np.int64(1992), np.int64(1993), np.int64(1994), np.int64(1995), np.int64(1996), np.int64(1997), np.int64(1998), np.int64(1999), np.int64(2000), np.int64(2001), np.int64(2002), np.int64(2003), np.int64(2004), np.int64(2005), np.int64(2006), np.int64(2007), np.int64(2008), np.int64(2009), np.int64(2010), np.int64(2011), np.int64(2012), np.int64(2013), np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022), np.int64(2023)]


In [40]:
print(*sorted(df['year'].unique()))

1972 1973 1986 1987 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019 2020 2021 2022 2023


### restart
seeing the huge difference in sampling methods I have decided to reload the data filtered on net sampling for Hypothesis 2.

In [47]:
# 1. Reload Config & Data
with open("../config.yaml", "r") as f:
    config = yaml.safe_load(f)
df = pd.read_parquet(config["output_data"]["file3"])

# 2. Preprocess
df['sample_date'] = pd.to_datetime(df['sample_date'], errors='coerce')
df['year'] = df['sample_date'].dt.year
# Ensure the measurement column is numeric
df['microplastics_measurement'] = pd.to_numeric(df['microplastics_measurement'], errors='coerce')
df = df.dropna(subset=['year', 'microplastics_measurement', 'lat', 'lng'])

# 3. Filter for NETS ONLY
net_keywords = [
    'Neuston net', 'Manta net', 'AVANI net', 'Plankton net', 
    'plankton net', 'manta net', 'net', 'Net'
]
pattern = '|'.join(net_keywords)
df_nets = df[df['sampling_method'].astype(str).str.contains(pattern, case=False, na=False)].copy()

print(f"✅ Filtered for Net Samples: {len(df_nets)} rows")

# 4. Select and Rename Columns for SQL
# We explicitly select the columns we want and ensure the name is 'microplastics_measurement'
df_sql = df_nets[['year', 'lat', 'lng', 'microplastics_measurement']].copy()

# 5. Upload to SQL

df_sql.to_sql(
    'observed_plastic_nets', 
    con=engine, 
    if_exists='replace', 
    index=False
)

print("\n✅ SUCCESS! Table 'observed_plastic_nets' created.")
print("   Columns: year, lat, lng, microplastics_measurement")
print("   You can now run your SQL queries using 'microplastics_measurement' instead of 'concentration'.")

✅ Filtered for Net Samples: 13385 rows

✅ SUCCESS! Table 'observed_plastic_nets' created.
   Columns: year, lat, lng, microplastics_measurement
   You can now run your SQL queries using 'microplastics_measurement' instead of 'concentration'.
